core/src/letter_distribution.c

```diff
@@ -114,23 +118,66 @@ void load_letter_distribution(LetterDistribution *letter_distribution,
 
     machine_letter++;
   }
-
+  letter_distribution->max_tile_length = max_tile_length;
   fclose(file);
 }
 
-// Assumes english
+// This is a linear search. This function should not be used for anything
+// that is speed-critical. If we ever need to use this in anything
+// speed-critical, we should use a hash.
 uint8_t
 human_readable_letter_to_machine_letter(LetterDistribution *letter_distribution,
-                                        unsigned char r) {
-  return (uint8_t)
-      letter_distribution->human_readable_letter_to_machine_letter[r];
+                                        char *letter) {
+
+  for (int i = 0; i < MACHINE_LETTER_MAX_VALUE; i++) {
+    if (strcmp(letter_distribution->machine_letter_to_human_readable_letter[i],
+               letter) == 0) {
+      return i;
+    }
+  }
+  // It better find it
+  return INVALID_LETTER;
+}
+
+void machine_letter_to_human_readable_letter(
+    LetterDistribution *letter_distribution, uint8_t ml,
+    char letter[MAX_LETTER_CHAR_LENGTH]) {
+  strcpy(letter,
+         letter_distribution->machine_letter_to_human_readable_letter[ml]);
 }
 
-unsigned char
-machine_letter_to_human_readable_letter(LetterDistribution *letter_distribution,
-                                        uint8_t ml) {
-  return (unsigned char)
-      letter_distribution->machine_letter_to_human_readable_letter[ml];
+// Convert a string of arbitrary characters into an array of machine letters,
+// returning the number of machine letters. This function does not allocate
+// the ml array; it is the caller's responsibility to make this array big
+// enough.
+// Note: This is a slow function that should not be used in any hot loops.
+int str_to_machine_letters(LetterDistribution *letter_distribution,
+                           const char *str, uint8_t *mls) {
+
+  int num_mls = 0;
+  int num_bytes = strlen(str);
+  int i = 0;
+  while (i < num_bytes) {
+    for (int j = i + letter_distribution->max_tile_length; j > i; j--) {
+      if (j > num_bytes) {
+        continue;
+      }
+      // possible letter goes from index i to j. Search for it.
+      char possible_letter[MAX_LETTER_CHAR_LENGTH];
+      memcpy(possible_letter, str + i, j - i);
+      possible_letter[j - i] = '\0';
+      uint8_t ml = human_readable_letter_to_machine_letter(letter_distribution,
+                                                           possible_letter);
+      if (ml == INVALID_LETTER) {
```

Comment:
Same as above.

Reply:
this is ok, as discussed in voice chat, it's an expected value.